### PR TITLE
[IMP]auth_totp: 2FA Trusted Devices

### PR DESCRIPTION
**Pull request by arg-odoo**
+ Added the trusted devices feature
+ Added remember this device checkbox on totp login page
+ Displayed list of trusted devices for odoo backend
+ Added conditional display for trusted devices on odoo backend (only if there is at least one device)
+ Displayed list of trusted devices for portal users with conditional display (same as above)
+ Added revoke all button for portal users and odoo backend
+ Removed trusted devices when disabling the two-factor authentication

Added '2FA Trusted Devices' feature to allow users to remember their device to bypass the 2FA for the next connections. The trusted devices are displayed in a 'Trusted Devices' list on the two-factor authentication settings page. It is possible to revoke the devices one by one or revoke all at once with a special button under the devices list. Changing the password or disabling two-factor authentication will revoke all trusted devices (except the one in use)

### CODE REVIEW

File: addons/auth_totp/views/user_preferences.xml [hunk 1]
```diff
@@ -130,11 +130,23 @@
...
+                                   title="What is this?" class="o_doc_link" target="_blank"/>
                             </span>
                             <button name="totp_disable" type="object" string="(Disable two-factor authentication)"
                                     class="btn btn-link text-muted"/>
                         </div>
+                        <div colspan="2" attrs="{'invisible': [('totp_trusted_device_ids', '=', [])]}">
+                            <field name="totp_trusted_device_ids" nolabel="1" colspan="4" readonly="1">
+                                <tree editable="bottom" create="false" delete="false">
+                                    <field name="name" string="Trusted Devices"/>
+                                    <field name="create_date" string="Memorization Date"/>
```
Comment (odony) [inline addons/auth_totp/views/user_preferences.xml:142]: simpler wording for users?
```suggestion
                                    <field name="create_date" string="Added On"/>
```

File: addons/auth_totp/views/user_preferences.xml [hunk 2]
```diff
@@ -130,11 +130,23 @@
...
+                                    <button type="object" name="remove" icon="fa-trash"/>
+                                </tree>
+                            </field>
+                            <button name="revoke_all_devices" string="Revoke All" type="object" class="btn btn-secondary" confirm="Are you sure you want
+                                to revoke all trusted devices ? You will have to pass the two-factor authentication again on every devices."/>
```
Comment (odony) [inline addons/auth_totp/views/user_preferences.xml:147]: more concise wording perhaps?

```suggestion
                            <button name="revoke_all_devices" string="Revoke All" type="object" class="btn btn-secondary" 
                                          confirm="Are you sure? Two-factor authentication will be required again on all your devices"/>
```

File: addons/auth_totp_portal/views/templates.xml [hunk 3]
```diff
@@ -41,6 +41,35 @@
                     <button type="button" class="btn btn-link" id="auth_totp_portal_disable">
                         (Disable two-factor authentication)
                     </button>
+                    <t t-if="len(user_id.totp_trusted_device_ids) > 0">
```
Comment (odony) [inline addons/auth_totp_portal/views/templates.xml:44]:
```suggestion
                    <t t-if="user_id.totp_trusted_device_ids">
```

**Review by odony — COMMENTED** [review]
Very nice work! :+1: 

Here's a first pass of comments, mostly minor things about naming conventions and wording.
I'll do a second pass after testing it properly, but it seems to be working well based on what I've seen so far!

I'm sure @xmo-odoo will be happy to comment as well.

When you're done with the fixes (and the tests?), could you squash your commits to prepare the branch for a clean merge? I imagine there will only be a single commit in the end, with the PR description as message.

Thanks!

File: addons/auth_totp/views/user_preferences.xml [hunk 4]
```diff
@@ -130,11 +130,23 @@
...
+                        </div>
+                        <span attrs="{'invisible': [('totp_trusted_device_ids', '!=', [])]}" class="text-muted">You don't have any recorded devices yet</span>
```
[the remark this answers is not in the document]
Comment (odony) [reply addons/auth_totp/views/user_preferences.xml:149]: > Or maybe leave this message but add a question mark icon with an on hover explanation like "Devices can be remembered when verifying two-factor authentication"? 

Hmm why not, what do you think @apr-odoo ?

File: odoo/addons/base/models/res_users.py [hunk 5]
```diff
@@ -1654,6 +1654,9 @@ def remove(self):
     def _remove(self):
         """Use the remove() method to remove an API Key. This method implement logic,
         but won't check the identity (mainly used to remove trusted devices)"""
+        if not len(self):
```
Comment (xmo-odoo) [inline odoo/addons/base/models/res_users.py:1657]: Just `not self` suffices.

Also not sure there's a point to returning something here? its `remove` doesn't use it, neither does `_revoke_all_devices`. Or is `remove` missing a `return`?

Comment (arg-odoo) [reply odoo/addons/base/models/res_users.py:1657]: Then `remove` should return the action dict too as it does before ?
So the code would be this one ?
```python
@check_identity
def remove(self):
    return self._remove()
```

Comment (xmo-odoo) [reply odoo/addons/base/models/res_users.py:1657]: If the intent is to return `close` when clicking on the button then yes, otherwise the return value is just lost. That's only useful if the button invoking `remove` is in a dialog though.

File: addons/auth_totp/static/tests/totp_flow.js [hunk 6]
```diff
@@ -243,5 +243,94 @@ trigger: 'div:contains("confirm your password")',
...
+}, {
+    content: 'Open user account menu',
+    trigger: '.o_user_menu .oe_topbar_name',
+    run: 'click',
+}, {
+    content: 'Log out',
+    trigger: 'a.dropdown-item',
+    run: 'click',
+},
+// Connect again to check if we can connect w/ totp
```
Comment (xmo-odoo) [inline addons/auth_totp/static/tests/totp_flow.js:296]: That should be `w/o` (for "without"), `w/` is "with".

**Review by KangOl — APPROVED** [review]
It seems functionally correct.
Just some small suggestions.

File: addons/auth_totp/models/res_users.py [hunk 7]
```diff
@@ -130,6 +137,18 @@ def totp_enable_wizard(self):
...
 
+    @check_identity
+    def revoke_all_devices(self):
+        self._revoke_all_devices()
+
+    def _revoke_all_devices(self):
+        self.totp_trusted_device_ids._remove()
+
+    def change_password(self, old_passwd, new_passwd):
+        self.env.user._revoke_all_devices()
```
[the remark this answers is not in the document]
Comment (odony) [reply addons/auth_totp/models/res_users.py:148]: Yes I think it's better to revoke trusted devices and force a real 2FA flow if anything unusual is happening with the account. Also FWIW, Arnaud conducted a survey of popular services with 2FA and confirmed that they also invalidate trusted devices as soon as the password is changed (e.g. Google or Stripe do that). This means it is a common practice, and one that users might expect.

**Comment by Xavier-Do** [comment]
Upgrade exception [60](https://runbot.odoo.com/web/#id=60&model=runbot.upgrade.exception&view_type=form&cids=1&menu_id=) added

**Comment by mart-e** [comment]
@robodoo r+
